### ISSUE TITLE
Custom icons must be in kebab case

### ISSUE THREAD

**Issue by huang-julien** [nuxt/icon#265]
## Description

iconify's `getIcon` uses `validateIconName` which enforce some validation for an icon's name. 

One validation is that a name must be split with `-` which is often not the case with users custom icons

## Reproduction 

https://stackblitz.com/edit/nuxt-starter-qlykyd?file=icons%2FEuro.svg,nuxt.config.ts,app.vue

in this reproduction, we can't see `my-icons:Euro` because Euro is in PascalCase

## Solution 

Either a documentation fix about naming convention or fix it in iconify or we could also merge prefix with the name ?

**Comment by YaredFall** [comment]
Stumbled upon this quirk too.

It seems that only lowercase letters, digits and dash (`-`) is allowed inside icon filename.

I was hoping to be able to use uppercase letters and underscores... I wonder why this "rule" is never mentioned in the docs.

**Comment by huang-julien** [comment]
Because we're using @iconify and this is the convention @iconify uses

**Comment by huang-julien** [comment]
I'll work on it when i have some free time

**Comment by polarove** [comment]
Same here, tortured me for an hour

**Comment by JonathanXDR** [comment]
Would it be possible to consider supporting dots (`.`) as separators for custom icons? This would add flexibility to naming conventions and align more closely with custom design systems, as, for example, it’s a common pattern at the company I work for.

**Comment by antfu** [comment]
Duplicated as #257

**Comment by cyberalien** [comment]
Naming convention is used in all parts of Iconify ecosystem, so changing it would be tricky. It would require way too many code changes.

I recommend changing it internally in Nuxt Icon. Convert names when loading and when passing icon name parameter. It should be as simple as changing icon name to lower case.

**Comment by cyberalien** [comment]
> In the long term, I do wish to support more relaxed conversion (at least `_` and `.`) - which I think might require us to wrapping or hack the iconify runtime a bit to do that.

I think support for any icon names can be added in future versions. Current convention will remain for icons loaded from API, but can be ignored for custom icons.

**Comment by cyberalien** [comment]
It is not simple. Icons are not loaded one by one, they are loaded in bulk from API, otherwise loading could take a while. So API end point is not easily customisable.

I think a simple solution would be to allow custom loaded based on prefix and/or provider. Then current API logic won't need to change, but it would allow custom solutions, which could be used to load icons from Nuxt.

**Comment by antfu** [comment]
> I think a simple solution would be to allow custom loaded based on prefix and/or provider.

Oh yeah, I think that would be great to serve our needs!

**Comment by cyberalien** [comment]
I'm thinking of making 3 ways to handle loading:
- `setIconSetLoader(callback, prefix, provider = '')` - loader for an entire icon set, could be useful if loading whole icon set at once.
- `setIconsLoader(callback, prefix, provider = '')` - loader for multiple icons, could be useful to fetch multiple icons in one request.
- `setIconLoader(callback, prefix, provider = '')` - loader for one icon, could be useful when fetching only one icon at a time.

Types for callbacks to make it clearer how callbacks could be implemented:
```ts
type IconSetLoaderCallback = (prefix: string, provider: string) => Promise<IconifyJSON | null>;
type IconsLoaderCallback = (names: string[], prefix: string, provider: string) => Promise<IconifyJSON | null>;
type IconLoaderCallback = (name: string, prefix: string, provider: string) => Promise<IconifyIcon | null>;
```

Idea is to set at least one method, icon component will use whatever is available: 
- if loading whole set is possible, use that
- if loading multiple icons is possible, use that
- otherwise load each icon one at a time

Parameters in function and callback are ordered from most useful first to least useful last. All loaders are async functions that return `Promise` instance. Result is data in correct type, `null` if icon is missing.

What do you think?

**Comment by cyberalien** [comment]
Working on solution: https://github.com/iconify/iconify/pull/348

**Comment by cyberalien** [comment]
Added 2 loaders:
- For multiple icons
- For one icon

You can use one or both. If both are set, loader for one icon is called when loading only one icon, otherwise loader for multiple icons is called.

Loaders have same signature, but different callbacks:
```js
setCustomIconsLoader(callback, prefix, provider = '')
setCustomIconLoader(callback, prefix, provider = '')
```
where `callback` is possibly async function that returns data or null.

Data is `IconifyJSON` icon set for `setCustomIconsLoader`, `IconifyIcon` icon for `setCustomIconLoader`.

Callback can be asynchronous or synchronous. Types for callbacks:
```ts
/**
 * Custom icons loader
 */
export type IconifyCustomIconsLoader = (
	icons: string[],
	prefix: string,
	provider: string
) => Promise<IconifyJSON | null> | IconifyJSON | null;

/**
 * Custom loader for one icon
 */
export type IconifyCustomIconLoader = (
	name: string,
	prefix: string,
	provider: string
) => Promise<IconifyIcon | null> | IconifyIcon | null;
```
Icon or list of icons is first parameter because other parameters are most likely will not be used, unless you are reusing same loader for multiple prefixes.

Why 2 loaders? Loader for one icon can be used as solution for customising icons: set loader for a custom prefix, use it to load icon from another icon set (possibly from API) using `loadIcon` function, modify response to change stuff like colors or stroke width, return modified icon:
```js

setCustomIconLoader(async (name) => {
	const data = await loadIcon(`tabler:${name}`);
	return data
		? {
				...data,
				body: data.body.replaceAll(
					'stroke-width="1.5"',
					'stroke-width="3"'
				),
		  }
		: null;
}, 'tabler-fat');
```

For now implemented in core, I'll add it to all components later today.

Loaders work with custom prefixes and icon names that have only one requirement: not being empty strings. So you can use whatever you want, even filenames (though `:` might cause problems, so strip that and trim string... basically do a simple sanity check in names to avoid surprises).

**Comment by cyberalien** [comment]
Published `@iconify/vue` version `4.2.0-dev.1` with new loader. Install as `@iconify/vue@next`.

**Comment by cyberalien** [comment]
One note about naming requirements. Icon still must have a prefix and name parts, separated with `:`. Separator cannot be changed and is required. Changing that behavior would require way too many changes, which is not feasible.

Names like `my-icons:Euro` now work correctly. Since other characters are not limited, for emulating file system stuff you can use `my-icons:svg/whatever.svg`, as long as there is only one `:` in name and prefix is set.

**Comment by cyberalien** [comment]
I think new component had enough testing, so published it as stable version `4.2.0` (minor version update).

**Comment by antfu** [comment]
Thanks a lot for the hard work @cyberalien, which consisted of such a long period of effort!

I made #317 to adopt to this feature
